File: JUtils/library/src/test/java/polanski/jutils/TestOption.java
package polanski.jutils;

import org.junit.Test;

import polanski.jutils.option.Option;
import rx.functions.Action0;
import rx.functions.Action1;
import rx.functions.Func0;
import rx.functions.Func1;
import rx.functions.Func2;
import rx.functions.Func3;
import rx.functions.Func4;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static polanski.jutils.option.Option.NONE;
import static polanski.jutils.option.Option.ofObj;
import static polanski.jutils.option.Option.tryAsOption;
import static polanski.jutils.option.OptionUnsafe.getUnsafe;

public class TestOption {

    @Test
    public void testOfObj_whenSome() {

        final String str = "Something";
        Option<String> op = ofObj(str);

        assertTrue(op.isSome());
        assertEquals(str, getUnsafe(op));
    }

    @Test
    public void testOfObj_whenNone() {

        final String str = null;
        Option<String> op = ofObj(str);

        assertFalse(op.isSome());
    }

    @Test
    public void testMap_whenSome() {

        final String str = "Something";
        Option<String> op = ofObj("")
                .map(new Func1<String, String>() {
                    @Override
                    public String call(final String __) {
                        return str;
                    }
                });

        assertTrue(op.isSome());
        assertEquals(str, getUnsafe(op));
    }

    @Test
    public void testMap_whenNone() {

        Option<String> op = ofObj((String) null)
                .map(new Func1<String, String>() {
                    @Override
                    public String call(final String __) {
                        return "";
                    }
                });

        assertFalse(op.isSome());
    }

    @Test
    public void testFilter_whenSome() {

        final String str = "Something";
        Option<String> op = ofObj(str)
                .filter(new Func1<String, Boolean>() {
                    @Override
                    public Boolean call(final String val) {
                        return val.equals(str);
                    }
                });

        assertTrue(op.isSome());
        assertEquals(str, getUnsafe(op));
    }

    @Test
    public void testFilterSome_whenFailed() {

        final String str = "Something";
        Option<String> op = ofObj(str)
                .filter(new Func1<String, Boolean>() {
                    @Override
                    public Boolean call(final String val) {
                        return val.equals("");
                    }
                });

        assertFalse(op.isSome());
    }

    @Test
    public void testFilter_whenNone() {

        Option<String> op = ofObj((String) null)
                .filter(new Func1<String, Boolean>() {
                    @Override
                    public Boolean call(final String val) {
                        return val.equals("");
                    }
                });

        assertFalse(op.isSome());
    }

    @Test
    public void testFlatMap_whenSome() {

        final String str = "Something";
        Option<String> op = ofObj("")
                .flatMap(new Func1<String, Option<String>>() {
                    @Override
                    public Option<String> call(final String val) {
                        return ofObj(str);
                    }
                });

        assertTrue(op.isSome());
        assertEquals(str, getUnsafe(op));
    }

    @Test
    public void testFlatMapSome_whenFailed() {

        final String str = "Something";
        Option<String> op = ofObj(str)
                .flatMap(new Func1<String, Option<String>>() {
                    @Override
                    public Option<String> call(final String val) {
                        return ofObj((String) null);
                    }
                });

        assertFalse(op.isSome());
    }

    @Test
    public void testFlatMap_whenNone() {

        Option<String> op = ofObj((String) null)
                .flatMap(new Func1<String, Option<String>>() {
                    @Override
                    public Option<String> call(final String val) {
                        return ofObj("");
                    }
                });

        assertFalse(op.isSome());
    }

    @Test
    public void testOrOption_whenSome() {

        final String str = "Something";
        Option<String> op = ofObj(str)
                .orOption(new Func0<Option<String>>() {
                    @Override
                    public Option<String> call() {
                        return ofObj("");
                    }
                });

        assertTrue(op.isSome());
        assertEquals(str, getUnsafe(op));
    }

    @Test
    public void testOrOption_whenNone() {
        final String str = "Something";
        Option<String> op = ofObj((String) null)
                .orOption(new Func0<Option<String>>() {
                    @Override
                    public Option<String> call() {
                        return ofObj(str);
                    }
                });

        assertTrue(op.isSome());
        assertEquals(str, getUnsafe(op));
    }

    @Test
    public void testOrDefault_whenSome() {

        final String str = "Something";
        String s = ofObj(str)
                .orDefault(new Func0<String>() {
                    @Override
                    public String call() {
                        return "";
                    }
                });

        assertEquals(str, s);
    }

    @Test
    public void testOrDefault_whenNone() {
        final String str = "Something";
        String s = ofObj((String) null)
                .orDefault(new Func0<String>() {
                    @Override
                    public String call() {
                        return str;
                    }
                });

        assertEquals(str, s);
    }

    @Test
    public void testTryAsOption_whenSome() {

        final String str = "Something";
        Option<String> op = tryAsOption(
                new Func0<String>() {
                    @Override
                    public String call() {
                        return str;
                    }
                });

        assertTrue(op.isSome());
        assertEquals(str, getUnsafe(op));
    }

    @SuppressWarnings("ConstantConditions")
    @Test
    public void testTryAsOption_whenNone() {
        final Integer str = null;

        //If converted to method reference then NPE is unavoidable
        @SuppressWarnings("Convert2MethodRef")
        Option<String> op = tryAsOption(
                new Func0<String>() {
                    @Override
                    public String call() {
                        return str.toString();
                    }
                });

        assertFalse(op.isSome());
    }

    @Test
    public void testMatchAction_whenSome() {
        final IFunction someFun = mock(IFunction.class);
        final IFunction noneFun = mock(IFunction.class);

        ofObj("").matchAction(new Action1<String>() {
                                  @Override
                                  public void call(final String it) {
                                      someFun.fun();
                                  }
                              },
                              new Action0() {
                                  @Override
                                  public void call() {
                                      noneFun.fun();
                                  }
                              });

        verify(someFun).fun();
        verify(noneFun, never()).fun();
    }

    @Test
    public void testMatchAction_whenNone() {
        final IFunction someFun = mock(IFunction.class);
        final IFunction noneFun = mock(IFunction.class);

        //noinspection unchecked
        NONE.matchAction(new Action1() {
                             @Override
                             public void call(final Object it) {
                                 someFun.fun();
                             }
                         },
                         new Action0() {
                             @Override
                             public void call() {
                                 noneFun.fun();
                             }
                         });

        verify(someFun, never()).fun();
        verify(noneFun).fun();
    }

    @Test
    public void testId() {
        final Option<String> op = ofObj("");

        assertEquals(op, op.id());
    }

    @Test
    public void testIfSome_whenSome() {
        final IFunction someFun = mock(IFunction.class);

        ofObj("").ifSome(new Action1<String>() {
            @Override
            public void call(final String it) {
                someFun.fun();
            }
        });

        verify(someFun, times(1)).fun();
    }

    @Test
    public void testIfSome_whenNone() {
        final IFunction noneFun = mock(IFunction.class);

        //noinspection unchecked
        NONE.ifSome(new Action1() {
            @Override
            public void call(final Object it) {
                noneFun.fun();
            }
        });

        verify(noneFun, never()).fun();
    }

    @Test
    public void testIfNone_whenSome() {
        final IFunction someFun = mock(IFunction.class);

        ofObj("").ifNone(new Action0() {
            @Override
            public void call() {
                someFun.fun();
            }
        });

        verify(someFun, never()).fun();
    }

    @Test
    public void testIfNone_whenNone() {
        final IFunction noneFun = mock(IFunction.class);

        //noinspection unchecked
        NONE.ifNone(new Action0() {
            @Override
            public void call() {
                noneFun.fun();
            }
        });

        verify(noneFun, times(1)).fun();
    }

    @Test
    public void testOfType_whenSomeAndProperType() {
        String value = "something";

        Option<String> op = ofObj((Object) value).ofType(String.class);

        assertTrue(op.isSome());
        assertEquals(value, getUnsafe(op));
    }

    @Test
    public void testOfType_whenSomeAndInvalidType() {
        String value = "something";

        Option<Integer> op = ofObj((Object) value).ofType(Integer.class);

        assertFalse(op.isSome());
    }

    @Test
    public void testOfType_whenNone() {
        //noinspection unchecked
        Option<String> op = NONE
                .ofType(String.class);

        assertFalse(op.isSome());
    }

    @Test
    public void testMatch_whenSome() {
        final String some = "some";
        final String none = "none";

        final String result = ofObj("").match(new Func1<String, String>() {
                                                  @Override
                                                  public String call(final String it) {
                                                      return some;
                                                  }
                                              },
                                              new Func0<String>() {
                                                  @Override
                                                  public String call() {
                                                      return none;
                                                  }
                                              });

        assertEquals(some, result);
    }

    @Test
    public void testMatch_whenNone() {
        final String some = "some";
        final String none = "none";

        //noinspection unchecked
        final String result = ofObj(null).match(new Func1<Object, String>() {
                                                    @Override
                                                    public String call(final Object it) {
                                                        return some;
                                                    }
                                                },
                                                new Func0<String>() {
                                                    @Override
                                                    public String call() {
                                                        return none;
                                                    }
                                                });

        assertEquals(none, result);
    }

    @Test
    public void testMatchUnsafe_whenSome() {
        final String some = "some";
        final String none = "none";

        final String result = ofObj("").matchUnsafe(new Func1<String, String>() {
                                                        @Override
                                                        public String call(final String it) {
                                                            return some;
                                                        }
                                                    },
                                                    new Func0<String>() {
                                                        @Override
                                                        public String call() {
                                                            return none;
                                                        }
                                                    });

        assertEquals(some, result);
    }

    @Test
    public void testMatchUnsafe_whenNone() {
        final String some = "some";
        final String none = "none";

        //noinspection unchecked
        final String result = ofObj(null).matchUnsafe(new Func1<Object, String>() {
                                                          @Override
                                                          public String call(final Object it) {
                                                              return some;
                                                          }
                                                      },
                                                      new Func0<String>() {
                                                          @Override
                                                          public String call() {
                                                              return none;
                                                          }
                                                      });

        assertEquals(none, result);
    }

    @Test
    public void testLift1_whenAllSome() {
        final Integer one = 1;
        final Integer two = 2;
        final Func2<Integer, Integer, Integer> fun = new Func2<Integer, Integer, Integer>() {
            @Override
            public Integer call(final Integer f, final Integer s) {
                return f + s;
            }
        };

        final Option<Integer> op = ofObj(one).lift(ofObj(two), fun);

        assertTrue(op.isSome());
        assertEquals(fun.call(one, two), getUnsafe(op));
    }

    @Test
    public void testLift1_whenFirstIsNone() {
        final Integer one = 1;
        final Func2<Integer, Integer, Integer> fun = new Func2<Integer, Integer, Integer>() {
            @Override
            public Integer call(final Integer f, final Integer s) {
                return f + s;
            }
        };

        //noinspection unchecked
        final Option<Integer> op = ofObj(one)
                .lift(NONE, fun);

        assertFalse(op.isSome());
    }

    @Test
    public void testLift1_whenSecondIsNone() {
        final Integer one = 1;
        final Func2<Integer, Integer, Integer> fun = new Func2<Integer, Integer, Integer>() {
            @Override
            public Integer call(final Integer f, final Integer s) {
                return f + s;
            }
        };

        //noinspection unchecked
        final Option<Integer> op = NONE
                .lift(ofObj(one), fun);

        assertFalse(op.isSome());
    }

    @Test
    public void testLift2_whenAllSome() {
        final Integer one = 1;
        final Integer two = 2;
        final Integer three = 3;
        final Func3<Integer, Integer, Integer, Integer> fun =
                new Func3<Integer, Integer, Integer, Integer>() {
                    @Override
                    public Integer call(final Integer f, final Integer s, final Integer t) {
                        return f + s + t;
                    }
                };

        final Option<Integer> op = ofObj(one)
                .lift(ofObj(two), ofObj(three), fun);

        assertTrue(op.isSome());
        assertEquals(fun.call(one, two, three), getUnsafe(op));
    }

    @Test
    public void testLift2_whenFirstIsNone() {
        final Integer one = 1;
        final Integer two = 2;
        final Func3<Integer, Integer, Integer, Integer> fun =
                new Func3<Integer, Integer, Integer, Integer>() {
                    @Override
                    public Integer call(final Integer f, final Integer s, final Integer t) {
                        return f + s + t;
                    }
                };

        //noinspection unchecked
        final Option<Integer> op = NONE
                .lift(ofObj(one), ofObj(two), fun);

        assertFalse(op.isSome());
    }

    @Test
    public void testLift2_whenSecondIsNone() {
        final Integer one = 1;
        final Integer two = 2;
        final Func3<Integer, Integer, Integer, Integer> fun =
                new Func3<Integer, Integer, Integer, Integer>() {
                    @Override
                    public Integer call(final Integer f, final Integer s, final Integer t) {
                        return f + s + t;
                    }
                };

        //noinspection unchecked
        final Option<Integer> op = ofObj(one)
                .lift(NONE, ofObj(two), fun);

        assertFalse(op.isSome());
    }

    @Test
    public void testLift2_whenThirdIsNone() {
        final Integer one = 1;
        final Integer two = 2;
        final Func3<Integer, Integer, Integer, Integer> fun =
                new Func3<Integer, Integer, Integer, Integer>() {
                    @Override
                    public Integer call(final Integer f, final Integer s, final Integer t) {
                        return f + s + t;
                    }
                };

        //noinspection unchecked
        final Option<Integer> op = ofObj(one)
                .lift(ofObj(two), NONE, fun);

        assertFalse(op.isSome());
    }

    @Test
    public void testLift3_whenAllSome() {
        final Integer one = 1;
        final Integer two = 2;
        final Integer three = 3;
        final Integer four = 4;
        final Func4<Integer, Integer, Integer, Integer, Integer> fun =
                new Func4<Integer, Integer, Integer, Integer, Integer>() {
                    @Override
                    public Integer call(final Integer f, final Integer s, final Integer t,
                                        final Integer fo) {
                        return f + s + t + fo;
                    }
                };

        final Option<Integer> op = ofObj(one)
                .lift(ofObj(two), ofObj(three), ofObj(four), fun);

        assertTrue(op.isSome());
        assertEquals(fun.call(one, two, three, four),
                     getUnsafe(op));
    }

    @Test
    public void testLift3_whenAnyIsNone() {
        final Integer one = 1;
        final Integer two = 2;
        final Integer three = 3;
        final Func4<Integer, Integer, Integer, Integer, Integer> fun =
                new Func4<Integer, Integer, Integer, Integer, Integer>() {
                    @Override
                    public Integer call(final Integer f, final Integer s, final Integer t,
                                        final Integer fo) {
                        return f + s + t + fo;
                    }
                };

        //noinspection unchecked
        final Option<Integer> op = ofObj(one)
                .lift(ofObj(two), NONE, ofObj(three), fun);

        assertFalse(op.isSome());
    }

    @Test
    public void testLift3_whenFirstIsNone() {
        final Integer one = 1;
        final Integer two = 2;
        final Integer three = 3;
        final Func4<Integer, Integer, Integer, Integer, Integer> fun =
                new Func4<Integer, Integer, Integer, Integer, Integer>() {
                    @Override
                    public Integer call(final Integer f, final Integer s, final Integer t,
                                        final Integer fo) {
                        return f + s + t + fo;
                    }
                };

        //noinspection unchecked
        final Option<Integer> op =
                NONE.lift(ofObj(one), ofObj(two), ofObj(three), fun);
        assertFalse(op.isSome());
    }

    @Test
    public void testToString_whenSome() {
        final Integer value = 1;

        final String result = ofObj(value).toString();

        assertEquals(value.toString(), result);
    }

    @Test
    public void testToString_whenNone() {
        final String result = NONE.toString();

        assertEquals(NONE.getClass().getSimpleName(), result);
    }

    @Test
    public void testHashCode_whenSome() {
        final Integer value = 1;

        final int result = ofObj(value).hashCode();

        assertEquals(value.hashCode(), result);
    }

    @Test
    public void testHashCode_whenNone() {
        final int result = NONE.hashCode();

        assertEquals(0, result);
    }

    @Test
    public void testOrThrowUnsafe_whenSome() {
        final int value = 1;

        final int result = polanski.jutils.option.OptionUnsafe
                .orThrowUnsafe(ofObj(value), new RuntimeException());

        assertEquals(value, result);
    }

    @Test(expected = RuntimeException.class)
    public void testOrThrowUnsafe_whenNone() {
        polanski.jutils.option.OptionUnsafe
                .orThrowUnsafe(NONE, new RuntimeException());
    }

    @Test
    public void testLog_usesLoggingFunction() {
        final IFunction loggingFun = mock(IFunction.class);

        ofObj("").log(new Action1<String>() {
            @Override
            public void call(final String str) {
                loggingFun.fun(str);
            }
        });

        verify(loggingFun, times(1)).fun(anyString());
    }

    @Test
    public void testLog_containsValueOfSome() {
        final IFunction loggingFun = mock(IFunction.class);
        Option<String> option = ofObj("something");

        option.log(new Action1<String>() {
            @Override
            public void call(final String str) {
                loggingFun.fun(str);
            }
        });

        verify(loggingFun, times(1)).fun(option.toString());
    }

    @Test
    public void testLog_containsNone() {
        //noinspection unchecked
        Option<String> option = NONE;
        final IFunction loggingFun = mock(IFunction.class);

        option.log(new Action1<String>() {
            @Override
            public void call(final String str) {
                loggingFun.fun(str);
            }
        });

        verify(loggingFun, times(1)).fun(option.toString());
    }

    @Test
    public void testLog_usesLoggingFunctionWithTag() {
        final IFunction loggingFun = mock(IFunction.class);

        ofObj("").log("", new Action1<String>() {
            @Override
            public void call(final String str) {
                loggingFun.fun(str);
            }
        });

        verify(loggingFun, times(1)).fun(anyString());
    }

    @Test
    public void testLog_usesTag() {
        final String tag = "someTag";

        ofObj("").log(tag, new Action1<String>() {
            @Override
            public void call(final String str) {
                assertTrue(str.contains(tag));
            }
        });
    }

    @Test
    public void testLog_withTag_containsValueOfSome() {
        final Option<String> option = ofObj("something");

        option.log("any", new Action1<String>() {
            @Override
            public void call(final String str) {
                assertTrue(str.contains(option.toString()));
            }
        });
    }

    interface IFunction {

        void fun();

        void fun(String str);
    }

}

File: options/core/src/test/java/polanski/option/TestOption.kt
package polanski.option

import org.assertj.core.api.Assertions.assertThat
import org.junit.Assert.*
import org.junit.Test
import org.mockito.Matchers.anyString
import org.mockito.Mockito.*
import polanski.option.Option.*
import polanski.option.OptionUnsafe.getUnsafe
import polanski.option.function.Func2
import polanski.option.function.Func3
import polanski.option.function.Func4

class TestOption {

    val fun2 = Func2<Int, Int, Int> { f, s -> f + s }
    val fun3 = Func3<Int, Int, Int, Int> { f, s, t -> f + s + t }
    val fun4 = Func4<Int, Int, Int, Int, Int> { f, s, t, fo -> f + s + t + fo }

    @Test
    fun testOfObj_whenSome() {
        val str = "Something"
        val op = ofObj(str)

        assertThat(op.isSome).isTrue()
        assertEquals(str, getUnsafe(op))
    }

    @Test
    fun testOfObj_whenNone() {
        val str: String? = null
        val op = ofObj(str)

        assertThat(op.isSome).isFalse()
    }

    @Test
    fun testIsSome_whenSome() {
        val op = ofObj("Something")

        assertThat(op.isSome).isTrue()
    }

    @Test
    fun testIsSome_whenNone() {
        val op = none<String>()

        assertThat(op.isSome).isFalse()
    }

    @Test
    fun testIsNone_whenSome() {
        val op = ofObj("Something")

        assertThat(op.isNone).isFalse()
    }

    @Test
    fun testIsNone_whenNone() {
        val op = none<String>()

        assertThat(op.isNone).isTrue()
    }

    @Test
    fun testMap_whenSome() {
        val str = "Something"
        val op = ofObj("").map { str }

        assertThat(op.isSome).isTrue()
        assertEquals(str, getUnsafe(op))
    }

    @Test
    fun testMap_whenNone() {
        val op = none<String>().map { "" }

        assertThat(op.isSome).isFalse()
    }

    @Test
    fun testFilter_whenSome() {
        val str = "Something"
        val op = ofObj(str).filter { it == str }

        assertThat(op.isSome).isTrue()
        assertEquals(str, getUnsafe(op))
    }

    @Test
    fun testFilterSome_whenFailed() {
        val str = "Something"
        val op = ofObj(str).filter { it == "" }

        assertThat(op.isSome).isFalse()
    }

    @Test
    fun testFilter_whenNone() {
        val op = none<String>().filter { it == "" }

        assertThat(op.isSome).isFalse()
    }

    @Test
    fun testFlatMap_whenSome() {
        val str = "Something"
        val op = ofObj("").flatMap { ofObj(str) }

        assertThat(op.isSome).isTrue()
        assertEquals(str, getUnsafe(op))
    }

    @Test
    fun testFlatMapSome_whenFailed() {
        val op = ofObj("Something").flatMap { none<String>() }

        assertThat(op.isSome).isFalse()
    }

    @Test
    fun testFlatMap_whenNone() {
        val op = none<String>().flatMap { ofObj("") }

        assertThat(op.isSome).isFalse()
    }

    @Test
    fun testOrOption_whenSome() {
        val str = "Something"
        val op = ofObj(str).orOption { ofObj("") }

        assertThat(op.isSome).isTrue()
        assertEquals(str, getUnsafe(op))
    }

    @Test
    fun testOrOption_whenNone() {
        val str = "Something"
        val op = none<String>().orOption { ofObj(str) }

        assertThat(op.isSome).isTrue()
        assertEquals(str, getUnsafe(op))
    }

    @Test
    fun testOrDefault_whenSome() {
        val str = "Something"
        val s = ofObj(str).orDefault { "" }

        assertEquals(str, s)
    }

    @Test
    fun testOrDefault_whenNone() {
        val str = "Something"
        val s = none<String>().orDefault { str }

        assertEquals(str, s)
    }

    @Test
    fun testTryAsOption_whenSome() {
        val str = "Something"
        val op = tryAsOption { str }

        assertThat(op.isSome).isTrue()
        assertEquals(str, getUnsafe(op))
    }

    @Test
    fun testTryAsOption_whenNone() {
        val str: Int? = null

        val op = tryAsOption { str!!.toString() }

        assertFalse(op.isSome)
    }

    @Test
    fun testMatchAction_whenSome() {
        val someFun = mock(IFunction::class.java)
        val noneFun = mock(IFunction::class.java)

        ofObj("").matchAction({ someFun.`fun`() })
        { noneFun.`fun`() }

        verify(someFun).`fun`()
        verify(noneFun, never()).`fun`()
    }

    @Test
    fun testMatchAction_whenNone() {
        val someFun = mock(IFunction::class.java)
        val noneFun = mock(IFunction::class.java)

        NONE.matchAction(
                { someFun.`fun`() }
                , { noneFun.`fun`() })

        verify(someFun, never()).`fun`()
        verify(noneFun).`fun`()
    }

    @Test
    fun testId() {
        val op = ofObj("")

        assertEquals(op, op.id())
    }

    @Test
    fun testIfSome_whenSome() {
        val someFun = mock(IFunction::class.java)

        ofObj("").ifSome { someFun.`fun`() }

        verify(someFun, times(1)).`fun`()
    }

    @Test
    fun testIfSome_whenNone() {
        val noneFun = mock(IFunction::class.java)

        NONE.ifSome { noneFun.`fun`() }


        verify(noneFun, never()).`fun`()
    }

    @Test
    fun testIfNone_whenSome() {
        val someFun = mock(IFunction::class.java)

        ofObj("").ifNone { someFun.`fun`() }

        verify(someFun, never()).`fun`()
    }

    @Test
    fun testIfNone_whenNone() {
        val noneFun = mock(IFunction::class.java)

        NONE.ifNone { noneFun.`fun`() }

        verify(noneFun, times(1)).`fun`()
    }

    @Test
    fun testOfType_whenSomeAndProperType() {
        val value = "something"

        val op = ofObj(value as Any).ofType(String::class.java)

        assertThat(op.isSome).isTrue()
        assertEquals(value, getUnsafe(op))
    }

    @Test
    fun testOfType_whenSomeAndInvalidType() {
        val value = "something"

        val op = ofObj(value as Any).ofType(Int::class.java)

        assertThat(op.isSome).isFalse()
    }

    @Test
    fun testOfType_whenNone() {
        val op = none<Int>().ofType(String::class.java)

        assertThat(op.isSome).isFalse()
    }

    @Test
    fun testMatch_whenSome() {
        val some = "some"
        val none = "none"

        val result = ofObj("").match({ some })
        { none }

        assertEquals(some, result)
    }

    @Test
    fun testMatch_whenNone() {
        val none = "none"

        val result = ofObj<Any>(null).match({ "some" })
        { none }

        assertEquals(none, result)
    }

    @Test
    fun testMatchUnsafe_whenSome() {
        val some = "some"

        val result = ofObj("").matchUnsafe({ some })
        { "none" }

        assertEquals(some, result)
    }

    @Test
    fun testMatchUnsafe_whenNone() {
        val none = "none"

        val result = ofObj<Any>(null).matchUnsafe({ "some" })
        { none }

        assertEquals(none, result)
    }

    @Test
    fun testLift1_whenAllSome() {
        val one = 1
        val two = 2

        val op = ofObj(one).lift(ofObj(two), fun2)

        assertThat(op.isSome).isTrue()
        assertEquals(fun2.call(one, two), getUnsafe(op))
    }

    @Test
    fun testLift1_whenFirstIsNone() {
        val op = ofObj(1).lift<Int, Int>(none(), fun2)

        assertFalse(op.isSome)
    }

    @Test
    fun testLift1_whenSecondIsNone() {
        val op = none<Int>().lift(ofObj(1), fun2)

        assertThat(op.isSome).isFalse()
    }

    @Test
    fun testLift2_whenAllSome() {
        val one = 1
        val two = 2
        val three = 3

        val op = ofObj(one).lift(ofObj(two), ofObj(three), fun3)

        assertThat(op.isSome).isTrue()
        assertEquals(fun3.call(one, two, three), getUnsafe(op))
    }

    @Test
    fun testLift2_whenFirstIsNone() {
        val op = none<Int>().lift(ofObj(1), ofObj(2), fun3)

        assertThat(op.isSome).isFalse()
    }

    @Test
    fun testLift2_whenSecondIsNone() {
        val op = ofObj(1).lift<Int, Int, Int>(none(), ofObj(2), fun3)

        assertThat(op.isSome).isFalse()
    }

    @Test
    fun testLift2_whenThirdIsNone() {
        val op = ofObj(1).lift<Int, Int, Int>(ofObj(2), none(), fun3)

        assertThat(op.isSome).isFalse()
    }

    @Test
    fun testLift3_whenAllSome() {
        val one = 1
        val two = 2
        val three = 3
        val four = 4

        val op = ofObj(one).lift(ofObj(two), ofObj(three), ofObj(four), fun4)

        assertThat(op.isSome).isTrue()
        assertEquals(fun4.call(one, two, three, four),
                getUnsafe(op))
    }

    @Test
    fun testLift3_whenAnyIsNone() {
        val op = ofObj(1).lift<Int, Int, Int, Int>(ofObj(2), none(), ofObj(3), fun4)

        assertThat(op.isSome).isFalse()
    }

    @Test
    fun testLift3_whenFirstIsNone() {
        val op = none<Int>().lift(ofObj(1), ofObj(2), ofObj(3), fun4)

        assertThat(op.isSome).isFalse()
    }

    @Test
    fun testLiftMany_whenFirstIsNone_returnNone() {
        val op = none<Int>().lift(1.rangeTo(4).map { ofObj(it) }, { it })

        assertThat(op.isSome).isFalse()
    }

    @Test
    fun testLiftMany_whenAnyIsNone_returnNone() {
        val op = ofObj(1).lift(2.rangeTo(4).map { ofObj(it) }.plus(none()), { it })

        assertThat(op.isSome).isFalse()
    }

    @Test
    fun testLiftMany_whenAllAreSome_returnSome() {
        val funN = polanski.option.function.FuncN<Int> {
            it.filterIsInstance<Int>()
                    .sum()
        }
        val first = 1
        val rest = 1..4
        val op = ofObj(first).lift(rest.map { ofObj(it) }, funN)

        assertThat(op.isSome).isTrue()
        assertEquals(rest.sum() + first, getUnsafe(op))
    }

    @Test
    fun testToString_whenSome() {
        val value = 1

        val result = ofObj(value).toString()

        assertEquals(value.toString(), result)
    }

    @Test
    fun testToString_whenNone() {
        val result = NONE.toString()

        assertEquals(NONE.javaClass.simpleName, result)
    }

    @Test
    fun testHashCode_whenSome() {
        val value = 1

        val result = ofObj(value).hashCode()

        assertEquals(value.hashCode().toLong(), result.toLong())
    }

    @Test
    fun testHashCode_whenNone() {
        val result = NONE.hashCode()

        assertEquals(0, result.toLong())
    }

    @Test
    fun testOrThrowUnsafe_whenSome() {
        val value = 1

        val result = OptionUnsafe.orThrowUnsafe(ofObj(value), RuntimeException())

        assertEquals(value.toLong(), result.toLong())
    }

    @Test(expected = RuntimeException::class)
    fun testOrThrowUnsafe_whenNone() {
        OptionUnsafe.orThrowUnsafe(NONE, RuntimeException())
    }

    @Test
    fun testLog_usesLoggingFunction() {
        val loggingFun = mock(IFunction::class.java)

        ofObj("").log { loggingFun.`fun`(it) }

        verify(loggingFun, times(1)).`fun`(anyString())
    }

    @Test
    fun testLog_containsValueOfSome() {
        val loggingFun = mock(IFunction::class.java)
        val option = ofObj("something")

        option.log { loggingFun.`fun`(it) }

        verify(loggingFun, times(1)).`fun`(option.toString())
    }

    @Test
    fun testLog_containsNone() {
        val option = NONE
        val loggingFun = mock(IFunction::class.java)

        option.log { loggingFun.`fun`(it) }

        verify(loggingFun, times(1)).`fun`(option.toString())
    }

    @Test
    fun testLog_usesLoggingFunctionWithTag() {
        val loggingFun = mock(IFunction::class.java)

        ofObj("").log("") { loggingFun.`fun`(it) }

        verify(loggingFun, times(1)).`fun`(anyString())
    }

    @Test
    fun testLog_usesTag() {
        val tag = "someTag"

        ofObj("").log(tag) { assertTrue(it.contains(tag)) }
    }

    @Test
    fun testLog_withTag_containsValueOfSome() {
        val option = ofObj("something")

        option.log("any") { assertTrue(it.contains(option.toString())) }
    }

    @Test
    fun testNone_returnsNONE() {
        assertEquals(Option.NONE, Option.none<Any>())
    }

    @Test(expected = IllegalStateException::class)
    fun testGetUnsafe_none() {
        Option.NONE.unsafe
    }

    @Test
    fun testGetUnsafe_some() {
        val value = 1

        assertEquals(value, ofObj(value).unsafe)
    }

    internal interface IFunction {

        fun `fun`()

        fun `fun`(str: String)
    }
}


